### ISSUE TITLE
Spurious wakeups in SyncResponseHandler

### ISSUE THREAD

**Issue by alja** [xrootd/xrootd#404]
Under extreme loads of the caching proxy (1000+ connections) occasionally crashes in XrdFileCache::IO::SetInput() as shown here:
https://gist.github.com/osschar/0ae4ca8d1e1cb1b89cd7f7dd9c1094d9#file-gistfile1-txt-L4441

Matevz and I think invalid read/writes in the SyncResponseHandler are due to spurious wakeups in the response handler. We have added protection against this by replacing semaphore with condition variable and do proper locking:
https://github.com/alja/xrootd/commit/2e4460393cf7f5615eb036924818a76dc76f21d0

With this change we repeated the helgrind test. The SyncResponseHandler errors were not reported anymore.

**Comment by osschar** [comment]
I was thinking about this some more ... the problem could also be that, in the original code, the section that writes into pStatus and pResponse is not under lock and thus their contents does not get synchronized with the thread waiting for the operation to complete. Then the cause for the behaviour might not be a spurious wakeup.

We can repeat the test by keeping the semaphore and introducing an additional mutex that will enforce the memory synchronization. And log spurious wakeups if we detect them.

**Comment by simonmichal** [comment]
Hi guys,
Thanks for reporting the problem!

It is rather suspicious that changing a semaphore to a condvar fixes the problem, after thinking about it, it occurred to me that we have a custom implementation of semaphores in xrootd because of:

https://sourceware.org/bugzilla/show_bug.cgi?id=12674
https://bugzilla.redhat.com/show_bug.cgi?id=1027348

now this problem has been fixed in glibc 2.21. Could you try to rebuild xrootd with a glibc 2.21 or newer and see if the problem persists?

There is a conditional compilation (https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClUglyHacks.hh#L32) and it will automatically use glibc semaphores.

Michal

**Comment by ljanyst** [comment]
I don't think it's spurious wake-ups because of how the Semaphore class is implemented in either case. See: https://github.com/xrootd/xrootd/blob/master/src/XrdSys/XrdSysPthread.hh#L345 and https://github.com/xrootd/xrootd/blob/master/src/XrdSys/XrdSysLinuxSemaphore.hh#L169

Personally, I would bet on cache coherency or optimization issues. I would recommend inserting a memory fence before the call to `pSem->Post()`. GCC has a builtin for this. Look for ` __sync_synchronize`.

**Comment by ljanyst** [comment]
I would also check if the problem disappears if you compile with: `-O0`. If it does disappear, then I would just do `asm volatile("": : :"memory")` instead of the full memory barrier. This has no runtime effect, it just makes GCC more careful with instruction reordering when doing optimizations.

**Comment by osschar** [comment]
Thanks Lukasz ... we will try this tomorrow then. Where do we insert the barrier, before
a) the if in https://github.com/xrootd/xrootd/blob/master/src/XrdSys/XrdSysPthread.hh#L341
or really before
b) https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClMessageUtils.hh#L66
? In case b) it will also be there for the libc semaphores that seem to be ok now.

With glibc-2.22 things work ok, the helgrind reports about SyncHandler are gone.

A question: why are semaphores better than condition variables in this case?

**Comment by ljanyst** [comment]
In light of this new evidence, I think we don't need the full memory barrier. Please insert `asm volatile("": : :"memory")` before line 66 and after line 90 in [XrdClMessageUtils.hh](https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClMessageUtils.hh). If this helps, I will create a nicer patch for the built-in semaphores.

The semaphores are better here because we just need a very trivial synchronization mechanism. Posting/waiting on a semaphore involves just one atomic update and one syscall in case of congestion. Condvars are far more complex and, typically, involve multiple locks and multiple updates.

You would typically use a condvar if you have:
1. a complex condition for signaling
2. need to bump the owning thread's priority either because the critical section (the mutex) has a priority ceiling or because you're blocking a thread of higher priority and the mutex is in the inheritance mode.

If you're interested in how a condvar may be implemented, take a look at this: [tb-condvar.c](https://github.com/ljanyst/thread-bites/blob/master/tb-condvar.c#L61). I was once bored and implemented pretty much all of the pthread functions NPTL-style mostly in simple C: https://github.com/ljanyst/thread-bites

**Comment by alja** [comment]
I added __sync_synchronize calls before semaphore Post() and after Wait():
https://github.com/alja/xrootd/commit/afac030d21a81de70a9ad4ccb044c69cb065c2ec

This did not fix the problem. There helgrind reports were the same.

**Comment by ljanyst** [comment]
Can you try the other one? It's interesting to me, so I'll try to figure out what makes the difference. If you need a quick fix, you can always use the condvar, but check if the cancellation works fine.

**Comment by alja** [comment]
There is no effect if added asm volatile("": : :"memory")  at the two places.

**Comment by osschar** [comment]
I don't understand how asm volatile could help if sync_synchronize did not ... that's why we tried sync_synchronize first and then didn't bother with the second one.

**Comment by ljanyst** [comment]
A gcc bug perhaps. We've seen these. But guessing seems counterproductive. I'll try to actually run this code. :)

**Comment by simonmichal** [comment]
From Valgrind manual (pragraph 7.5, http://valgrind.org/docs/manual/hg-manual.html):

> Make sure your application, and all the libraries it uses, use the POSIX threading primitives. 
> Helgrind needs to be able to see all events pertaining to thread creation, exit, locking and other 
> synchronisation events. To do so it intercepts many POSIX pthreads functions.

> Do not roll your own threading primitives (mutexes, etc) from combinations of the Linux futex 
> syscall, atomic counters, etc. These throw Helgrind's internal what's-going-on models way off 
> course and will give bogus results.

The point is, even if our custom implementation of semaphores is correct, hellgring wont be able to tell ;-)

@alja and @osschar did you guys stress tested the caching proxy with condvar or just checked with hellgrind? if its just hellgrind maybe it would be worth to do the stress test to see if the crashes in XrdFileCache::IO::SetInput() really go away?

**Comment by ljanyst** [comment]
Good catch!

All this is incredibly interconnected. I have seen in glibc code to suppress Helgrind warnings in the cases where they couldn't use standard synchronization methods.

**Comment by osschar** [comment]
We are having it stress tested by the guys at Notre Dame where we initially saw those crashes, typically once per day when the proxies were under heavy load. I'll check things are going there, they updated on 8/19.

**Comment by abh3** [comment]
Can we get an update on this? We'd like to close it, if possible.

**Comment by osschar** [comment]
Yes, I'd like to close it, too :)

The workaround using condition variables (instead of custom written semaphores) has proved to be stable ... it has been running without crashes for a month now at a site where they saw it crash daily before.

I could live with run-time configuration that can be flipped from the caching proxy initialization code. But its' going to make code there a bit messier.

What do you propose?

**Comment by xrootd-dev** [comment]
If it's working and you are  happy with it thenkeep it.

**Comment by abh3** [comment]
Happy resolution.

**Comment by osschar** [comment]
OK, then I'll make a PR with this change for the pssasync branch ... and then it gets merged into the master.

Thanks!
